mlflow/gateway/schemas/chat.py

```diff
@@ -1,17 +1,119 @@
-from typing import Literal, Optional
+from typing import Literal, Optional, Union
 
-from pydantic import Field
+from pydantic import BaseModel, Field
+from typing_extensions import Annotated
 
 from mlflow.gateway.base_models import RequestModel, ResponseModel
-from mlflow.gateway.config import IS_PYDANTIC_V2
+from mlflow.utils import IS_PYDANTIC_V2
 
 
-class RequestMessage(RequestModel):
+class TextContentPart(RequestModel):
+    type: Literal["text"]
+    text: str
+
+
+class ImageUrl(RequestModel):
+    url: str  # either URL of an image, or bas64 encoded data
+    detail: Literal["auto", "low", "high"]
+
+
+class ImageContentPart(RequestModel):
+    type: Literal["image_url"]
+    image_url: ImageUrl
+
+
+class InputAudio(RequestModel):
+    data: str  # base64 encoded data
+    format: Literal["wav", "mp3"]
+
+
+class AudioContentPart(RequestModel):
+    type: Literal["input_audio"]
+    input_audio: InputAudio
+
+
+ContentPartsList = Annotated[
+    list[
+        Annotated[
+            Union[TextContentPart, ImageContentPart, AudioContentPart], Field(discriminator="type")
+        ]
+    ],
+    Field(min_items=1),
+]
+
+
+ContentType = Annotated[Union[str, ContentPartsList], Field(union_mode="left_to_right")]
+
+
+class Function(ResponseModel):
+    name: str
+    arguments: str
+
+
+class ToolCall(ResponseModel):
+    id: str
+    type: Literal["function"]
+    function: Function
+
+
+class RequestMessage(BaseModel):
+    """
+    A chat request. ``content`` can be a string, or an array of content parts.
+
+    A content part is one of the following:
+
+    - :py:class:`TextContentPart <mlflow.gateway.schemas.chat.TextContentPart>`
+    - :py:class:`ImageContentPart <mlflow.gateway.schemas.chat.ImageContentPart>`
+    - :py:class:`AudioContentPart <mlflow.gateway.schemas.chat.AudioContentPart>`
+    """
+
     role: str
-    content: str
+    content: Optional[ContentType] = None
+    tool_calls: Optional[list[ToolCall]] = Field(None, min_items=1)
+    tool_call_id: Optional[str] = None
+    refusal: Optional[str] = None
```

Comment:
i wish there was a way to simply define a type union like

```
RequestMessage = Union[UserMessage, SystemMessage, AssistantMessage, ToolMessage]
```

but it seems like every way of doing this has some problems. If we simply define it as a type (see Cat and Pet [in this example](https://docs.pydantic.dev/latest/concepts/unions/#nested-discriminated-unions)), then we can't do `type.model_validate()`, because it is not a class.

if we [extend RootModel](https://docs.pydantic.dev/latest/concepts/unions/#nested-discriminated-unions), then we can't access fields like `message.role` (and this RootModel class does not exist in Pydantic 1)

Reply:
I agree with this approach and indeed don't think we wanna separate subclasses for types. The goal is to be OpenAI compatible and they define [the schema](https://github.com/openai/openai-python/blob/e94d98e9bf97a5d2d02d79d58f2abdbab26ff2bd/src/openai/types/chat/chat_completion_message.py#L26) as a unified chat message class. The design of having subclasses per type seems to be Langchain-unique.